Enrollment technical interfaces

christiaanps opened the pull request.

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,267 @@
+openapi: 3.0.0
+info:
+  description: OSIA Enrollment Interface
+  version: 1.1.0
```
Should be version 1.0.0

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,267 @@
+                $ref: '#/components/schemas/Error'
+    put:
+      tags:
+        - enrollment
```
Typo: should be "Enrollment" (uppercase)

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,267 @@
+                $ref: '#/components/schemas/Error'
+
+  /v1/enrollments:
+    get:
```
Passing a list of expressions as query parameters might be cumbersome. I propose to pass this list in the body, as done for [DataAccess/verify](https://osia.readthedocs.io/en/latest/annexes/technical/dataaccess.html#post--v1-persons-uin-verify). This means we would have to use a POST (and not a GET).

petr-sturc-thales replied:
With regards to the expression in body. Was it said how they are evaluated in case of multiple expressions?  Logical AND or OR?

christiaanps replied:
> With regards to the expression in body. Was it said how they are evaluated in case of multiple expressions? Logical AND or OR?
Good question. I would think maybe we could use regular expression?

olivier-heurtier-idemia commented on src/doc/functional/_enrollment.rst:
```diff
@@ -7,119 +7,90 @@ Enrollment Services
     :return: a status indicating success or error and in case of success the enrollment data.
 
-.. py:function:: updateEnrollment(enrollmentID, enrollmentData, transactionID)
+.. py:function:: updateEnrollment(enrollmentID, enrollmentTypeId, enrollmentFlags, requestData, biometricData, biographicData, documentData, transactionID)
```
Could you also add a "partialUpdateEnrollment" service (I think we agreed we needed both, like in PR interface)

olivier-heurtier-idemia commented on src/doc/functional/_enrollment.rst:
```diff
@@ -7,119 +7,90 @@ Enrollment Services
-
-.. py:function:: createDocumentCapture(enrollmentID, documentID, documentData, transactionID)
+	
+.. py:function:: finalizeEnrollments(enrollmentID, transactionID)
```
Rename as "finalizeEnrollment" (no 's')

olivier-heurtier-idemia commented:
Hi @christiaanps 
I reviewed the functional description and added a few comments inline. Globally I'm fine with it.
Do you think it would be possible to add in all service an optional "finalize" flag, as discussed during the workshops?
Thanks

olivier-heurtier-idemia commented on src/doc/functional/_enrollment.rst:
```diff
@@ -2,124 +2,141 @@
 -------------------
 
+This interface describes enrollment services in the context of an identity system. It is based on
+the following principles:
+- When enrollment is done in one step, the CreateEnrollment can contain all the data and an additional flag (finalize) to indicate every data was collected.
```
Typo: missing empty line, needed by Sphinx/rst to recognize a bullet list is next.

olivier-heurtier-idemia commented on src/doc/functional/_enrollment.rst:
```diff
@@ -2,124 +2,141 @@
+    :return: a status indicating success or error and in case of success the buffer ID.
 
-.. py:function:: deleteDocumentCapture(documentID, transactionID)
+.. py:function:: getBuffer(enrollmentId, bufferId, data)
```
No data in getBuffer inputs

olivier-heurtier-idemia commented on src/doc/functional/_enrollment.rst:
```diff
@@ -2,124 +2,141 @@
 
-    :param str documentID: The ID of the document.
+    :param str enrollmentID: The ID of the enrollment.
+	:param str bufferID: The ID of the buffer.
```
Typo: bad indent

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+        - name: finalize
+          in: query
+          description: Flag to indicate that data was collected
+          required: true
```
Should be optional maybe? (default to false)

christiaanps replied:
> Should be optional maybe? (default to false)
what is the property for default?

olivier-heurtier-idemia replied:
I don't think we can define the default value in OpenAPI. This would have to be added to the description.

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+          schema:
+            type: string
+        - name: attributes
+          in: query
+          description: The (optional) set of required attributes to retrieve
+          required: false
+          schema:
+            type: array
+            items:
+              type: string
+            example:
+              - surname
+              - familyname
```
Can we assume that if not present all attributes are returned? If yes, can you add this in the description?

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+      summary: Retrieve a list of enrollments which match passed in search criteria
+      operationId: findEnrollments
+      parameters:
+        - name: expressions
+          in: query
+          description: The expressions to evaluate.
+          required: true
+          schema:
+            $ref: '#/components/schemas/Expressions'
```
Expressions are both in query and in body. I think this 'query' parameter should be removed.

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+        description: The image of the request
+        required: true
+        content:
+          plain/text:
+            schema:
+              type: string
+              format: byte
```
We should favour binary transfer using type application/octet-stream and/or image/jpeg and/or ... Format should be "binary".
See https://swagger.io/docs/specification/describing-request-body/file-upload/ for more info.

christiaanps replied:
I will go for  application/octet-stream

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+                required:
+                  - bufferId
+                properties:
+                  enrollmentId:
```
typo: bufferId

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+        - enrollmentType
+        - biometricData
+      properties:
+        encounterId:
```
typo: enrollmentId

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+          type: string
+        status:
+          type: string
+          enum: [ACTIVE, INACTIVE]
```
In the context of the enrollment, I suggest "FINALIZED" or "IN_PROGRESS" (or something equivalent).

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,652 @@
+        enrollmentType:
+          type: string
+          description: Type of the enrollment
+        enrollmentFlags:
+          type: array
+          items:
+            type: string
+          minItems: 1
+          uniqueItems: true
+        requestData:
+          type: array
+          items:
+            type: string
+          minItems: 1
+          uniqueItems: true
```
Should be a dictionary and not a list to be aligned on the functional description of chapter 5.4.4

christiaanps replied:
RequestData can be a disctionary but enrollmentFlags should be a list? Or should I make it a dictionary?

christiaanps replied:
Will fix this

olivier-heurtier-idemia commented:
Hi @christiaanps 
I reviewed your last changes and the YAML. See my comments above plus those 2 general comments:
- I'm not sure but I think we agreed to add a "mimetype" field in biometric and document structure to keep track of the format of the buffer.
- Can you also add the corresponding subchapter in 7.3 (Annexes). If you don't feel comfortable with that you can leave as it is now and I will add it later on.

Thanks for your contribution, result is excellent.

christiaanps commented:
Hi @olivier-heurtier-idemia 

I have fixed the PR. Also I have added the mimetype and added the Annexes but I think it needs to be referenced somewhere?

Thank you for the good review!

olivier-heurtier-idemia commented on src/doc/yaml/enrollment.yaml:
```diff
@@ -0,0 +1,666 @@
+        - LIVE_SCAN_OPTICAL_CONTACTLESS_PLAIN
+        - OTHER
+        - UNKNOWN
+    DocumentType:
+      type: string
+      enum:
+        - ID_CARD
+        - Passport
```
Could you sync with the PR interface which has:
```
      type: string
      description: Type of document. Can be extended.
      enum: [ID_CARD, PASSPORT]
```
What is important is to indicate this list can be extended.

olivier-heurtier-idemia commented:
After discussing internally in IDEMIA, we think that it would be interesting to add a DeleteEnrollment.
It is not very complex and can allow a simple & proper management of resources for those that do not trust completely a GC mechanism.

olivier-heurtier-idemia commented:
For consistency with the other API (PR & ABIS mostly), could you pluralize the resource name in the path (/v1/enrollment -> /v1/enrollment**s**)
This seems to be the usual convention (See https://opensource.zalando.com/restful-api-guidelines/#134 or the Google calendar API for example).

christiaanps commented:
> For consistency with the other API (PR & ABIS mostly), could you pluralize the resource name in the path (/v1/enrollment -> /v1/enrollment**s**)
> This seems to be the usual convention (See https://opensource.zalando.com/restful-api-guidelines/#134 or the Google calendar API for example).

I already have v1-enrollments to find enrollments. How would you like me to change this path? Should I add   /v1/enrollments/find?

olivier-heurtier-idemia commented:
> 
> 
> > For consistency with the other API (PR & ABIS mostly), could you pluralize the resource name in the path (/v1/enrollment -> /v1/enrollment**s**)
> > This seems to be the usual convention (See https://opensource.zalando.com/restful-api-guidelines/#134 or the Google calendar API for example).
> 
> I already have v1-enrollments to find enrollments. How would you like me to change this path? Should I add /v1/enrollments/find?

Hi @christiaanps 
Since there is no conflict on the paths you can just add a 's' to all the path and keep '/v1/enrollments' for the find enrollments service.

christiaanps commented:
> > > For consistency with the other API (PR & ABIS mostly), could you pluralize the resource name in the path (/v1/enrollment -> /v1/enrollment**s**)
> > > This seems to be the usual convention (See https://opensource.zalando.com/restful-api-guidelines/#134 or the Google calendar API for example).
> > 
> > 
> > I already have v1-enrollments to find enrollments. How would you like me to change this path? Should I add /v1/enrollments/find?
> 
> Hi @christiaanps
> Since there is no conflict on the paths you can just add a 's' to all the path and keep '/v1/enrollments' for the find enrollments service.

I have updated the PR
